[RemoteMirror] Fix lookup of generic type parameters

tbkka opened the pull request:
PR #62854 identified an issue with generic type arguments.  Unfortunately, that fix didn't quite work, since it inadvertently renumbered those arguments, leading to failures when we tried to look up generic type arguments when we had nested generic and non-generic types interspersed.

This is (I hope) a more correct fix.  It changes how the parent types are recursively reconstructed to ensure that we number the nesting levels the same way the compiler does.  I've also built some additional tests that verify not only the type reference (name) is correctly printed, but that the built types have the correct layout (which implies that the various generic type arguments are being correctly handled).

Resolves rdar://106790077

tbkka committed:
Partially Revert "Fix wrong calculation of generic params per level when building decl"
(Keeps the test, but reverts the code changes; the test will be updated later.)

This reverts most of commit 893d83ba0ad8ae279d2f7c792a237618b54476bd from PR #62854

tbkka committed:
Recursively reconstruct nested generics, skipping non-generic

This is a more correct fix for the issue that inspired PR #62854.
In particular, this does not change the numbering of generic
argument levels, and so does not break generic type parameter
lookups.

Added a new test case to verify that nested types that mix
generics and non-generics in different ways consistently identify
the correct generic argument labels.

tbkka committed:
Correct and simplify the test from PR #62854

The new logic no longer includes non-generic parent types as part
of the type reference.

This also combines the 32- and 64-bit tests validations, since they
were identical for this test anyway.

tbkka commented on include/swift/Remote/MetadataReader.h:
```diff
@@ -2959,19 +2959,13 @@ class MetadataReader {
-            } else {
-              // If there is no generic context, this is a non-generic type
-              // which has 0 generic parameters.
-              paramsPerLevel.emplace_back(0);
```
This was the key bug introduced by PR #62854: Adding a `0` params marker here caused non-generic types to get counted as generic types, throwing off the generic type parameter matching.

tbkka commented on include/swift/RemoteInspection/TypeRefBuilder.h:
```diff
@@ -567,62 +567,58 @@ class TypeRefBuilder {
   }
 
+  // Recursively constructs TypeRef for a bound generic, including
+  // the enclosing (parent) generic contexts
```
I've significantly rearranged the logic here to make it more cleanly recursive (which allowed me to eliminate the iterative walk to find a parent type).

tbkka commented on validation-test/Reflection/reflect_nested.swift:
```diff
@@ -1,7 +1,7 @@
 // RUN: %target-build-swift -lswiftSwiftReflectionTest %s -o %t/reflect_nested
 // RUN: %target-codesign %t/reflect_nested
-// RUN: %target-run %target-swift-reflection-test %t/reflect_nested 2>&1 | %FileCheck %s --check-prefix=CHECK-%target-ptrsize
+// RUN: %target-run %target-swift-reflection-test %t/reflect_nested 2>&1 | %FileCheck %s --check-prefix=CHECK
```
Since the CHECK-64 and CHECK-32 lines here are all identical, I just eliminated the target-ptrsize distinction.

tbkka commented on validation-test/Reflection/reflect_nested.swift:
```diff
@@ -26,24 +26,14 @@ var obj = OuterGeneric.Inner.Innermost(x: 17, y: "hello")
-// CHECK-64: Type reference:
-// CHECK-64: (bound_generic_class reflect_nested.OuterGeneric.Inner.Innermost
-// CHECK-64-NEXT:   (struct Swift.String)
-// CHECK-64-NEXT:   (bound_generic_class reflect_nested.OuterGeneric.Inner
```
The one actual change:  reflect_nested.OuterGeneric.Inner is not a bound_generic_class, so does not need to be in the list of generic parent types.

tbkka commented on validation-test/Reflection/reflect_nested_generic.swift:
```diff
@@ -0,0 +1,117 @@
+struct Outer2 {
+  enum E<T: P> {
+    struct Inner {
+      enum F<U: P> {
```
Changing this to a multi-payload Enum fails.  I think that's just because all the MPE fixes from #64852 haven't landed yet.

tbkka reviewed (commented):
A few comments on the code here to help reviewers.

tbkka commented:
@swift-ci Please test

mikeash reviewed (approved):
Nice, that's a subtle bug.

augusto2112 commented on include/swift/RemoteInspection/TypeRefBuilder.h:
```diff
@@ -567,62 +567,58 @@ class TypeRefBuilder {
+
+    // Nodes with no generic args can be replaced with their parent
+    if (numGenericArgs == 0) {
+      return createBoundGenericTypeReconstructingParent(
```
I don't understand the logic here, why can we ignore the current node? From the lldb test failure, it looks like we're dropping the innermost non-generic type, maybe because of this?

Given:
```
class F {
  class G<T> {
    class H {
      ...
    }
  }
}
```

The typeref of `F.G<Bool>.H` is being reconstructed as F.G<Bool>

augusto2112 replied:
I think it's because a non-generic nested inside a generic one is actually generic, but not the other way around (which was my original mistake)?

So:

```
class F { // F is not generic.
  class G<T> { // G is generic.
    class H { // H is generic.
      ...
    }
  }
}
```

tbkka replied:
Oh, good catch.  Yeah, I mixed something up here.  There's a basic difference between the original type we're constructing a reference for (which we have to reconstruct even if it's not a bound generic) and the parent types (where we only care about the bound generics).  That difference isn't being reflected in the code here.

The tricky part:  I don't think "bound generic" can mean "relies on generic information."  It has to mean "has generic type parameters."   So in your code above, we should not treat `H` as a "bound generic."  (Otherwise, `H` is ambiguous; we'd have to treat it as a bound generic when evaluated directly, but not when we evaluate another generic type contained inside it.)  But you're right that we have to construct a TypeRef for the type we're currently examining regardless of whether it's a bound generic or not.

tbkka replied:
I think I've now correctly reworked things to handle this case correctly as well.  Basically, we now include the innermost typeref regardless (since that's the type we're actually trying to build a typeref for), but only include parent nodes that have non-zero numbers of generic type parameters.  I also added a test to check the case where the innermost type has no generic parameters.

tbkka committed:
Add a test case where the innermost type has no generic parameters

tbkka committed:
Rework createBoundGenericTypeReconstructingParent

The recursive structure made it a little awkward to correctly
distinguish between the root node (which has to be included
regardless of whether it has direct generic params)
and the parent nodes (which must only be included if they
have direct params).

So I rewrote this to do a simple two-pass iteration:
* The first pass walks the parent list collecting candidates
* The second pass walks the list backwards, assigning generic params

We then just stack the start node onto the end of the
list regardless of whether it has generic params.

tbkka committed:
Refactor to handle the parent list separately

Handle the top node directly in `createBoundGenericType`
so that we can be sure to always include it regardless
of whether it has direct generic type params or not.

The helper function can now focus only on handling
parents (where we must only include ones that actually
contain generic type params).  This separates the two
different cases and makes it a little easier to follow.

I refactored `reconstructParentsOfBoundGenericType` to
be iterative rather than recursive, though I'm not sure that
really matters.

tbkka commented on include/swift/RemoteInspection/TypeRefBuilder.h:
```diff
@@ -567,65 +567,66 @@ class TypeRefBuilder {
+  const BoundGenericTypeRef *reconstructParentsOfBoundGenericType(
+      const NodePointer startNode,
+      const std::vector<size_t> &genericParamsPerLevel,
+      const llvm::ArrayRef<const TypeRef *> &args)
```
After fussing around with different structures here, I finally reduced this to _only_ handle reconstructing the parent list.  That means it can consistently ignore nodes with no direct generic params.  I also reworked it to be iterative rather than recursive, though that's not particularly relevant -- I just found it a little easier to follow that way.

tbkka commented on include/swift/RemoteInspection/TypeRefBuilder.h:
```diff
@@ -567,65 +567,66 @@ class TypeRefBuilder {
+    // We're now going to build the type tree from the
+    // outermost parent in, which matches the order of
+    // the generic parameter list and genericParamsPerLevel.
+    std::reverse(nodes.begin(), nodes.end());
```
Above, we walk the parent list from the innermost to outermost.  That is, for `Foo<Int>.Bar<Double>.Baz<String>.Quux`, we visit `Baz`, then `Bar`, then `Foo` (skipping `Quux`, since that's the start node, not one of the parents we're collecting here.

I then need to build the new list in the opposite order, which is why I reverse the list here.

tbkka commented on include/swift/RemoteInspection/TypeRefBuilder.h:
```diff
@@ -567,65 +567,66 @@ class TypeRefBuilder {
+      }
+      auto argEnd = argBegin + numGenericArgs;
+      std::vector<const TypeRef *> params(argBegin, argEnd);
+      argBegin = argEnd;
```
Walking the list in order makes it a lot easier to track which arguments go with with parent node.

tbkka commented on include/swift/RemoteInspection/TypeRefBuilder.h:
```diff
@@ -634,17 +635,58 @@ class TypeRefBuilder {
+    // Sanity:  Verify that the generic params per level add
+    // up exactly to the number of args we were provided, and
+    // that we don't have a rediculous number of either one
+    auto genericParamsPerLevel = *maybeGenericParamsPerLevel;
```
There were a lot of checks to ensure that the shape list (params-per-level) and the actual args list made sense together.  I found it easier to pull that together into a single unified check:  Ensure that the total number of params mentioned in params-per-level adds up exactly to the number of args we have.  I also imposed an arbitrary limit of 1000 nested types and 1000 generic args in order to be comfortably certain that we'll never have problems from integer wrapping.

tbkka commented on include/swift/RemoteInspection/TypeRefBuilder.h:
```diff
@@ -634,17 +635,58 @@ class TypeRefBuilder {
+    std::vector<const TypeRef *> params(argBegin, args.end());
+
+    // Build and return the top typeref
+    return BoundGenericTypeRef::create(*this, mangling.result(), params, parents);
```
The innermost node always gets included here, since that's the node we're actually constructing the typeref for.  This happens even if the innermost node has no explicit generic type params.

tbkka commented on validation-test/Reflection/reflect_nested_generic.swift:
```diff
@@ -0,0 +1,155 @@
+//CHECK: Enum value:
+//CHECK-NEXT: (enum_value name=b index=2)
+
+reflect(enum: Outer2.E<S1>.Inner.F<S2>.Innerer?.none)
```
As @augusto2112 suggested, I added a test where the innermost type has no generic type parameters.

tbkka commented:
@swift-ci Please test

augusto2112 reviewed (approved):
This looks much better than the original!

mikeash commented on include/swift/RemoteInspection/TypeRefBuilder.h:
```diff
@@ -567,65 +567,66 @@ class TypeRefBuilder {
+    // Collect the first N parents that potentially have generic args
+    // (Ignore the last genericParamPerLevel, which
+    // applies to the startNode itself.)
+    std::vector<NodePointer> nodes;
```
Fairly unimportant, but if you use a `SmallVector` here, it'll avoid allocating in the common case where there number of parents is small.
